Review data model (#651)
unicode-org/message-format-wg: spec/data-model/README.md

## Messages

A `SimpleMessage` corresponds to a syntax message that includes _selectors_.

> A message without _selectors_ and with a single _pattern_ is represented by a `ComplexMessage`.

Reviewer: This one is backwards.

```suggestion
A `SimpleMessage` is a _message_ without _selectors_ and with a single _pattern_.
A `ComplexMessage` is any _message_ that includes one or more _declarations_, a _selector_, or both.

In the syntax,

> a `ComplexMessage` may be represented either as a _simple message_ or as a _complex message_,

Reviewer 2: I think it's confusing to use "complex" to mean two different things (in the data model, having only a single pattern, regardless of declarations; in the concrete syntax, having no declarations and no selectors.)

I don't love the terms `PatternMessage` and `SelectMessage`, but I agree with Eemeli that it's probably best to keep it that way.

depending on whether it has declarations and if its `pattern` is allowed in a _simple message_.

```ts

> type Message = SimpleMessage | ComplexMessage;

Reviewer 3: Why this change in representation? This seems wholly unnecessary, as it links the data model with syntactic details like the representation of a message with a single pattern depending on whether it starts with a `.` or not.

I would have expected a change like this to have at least some argumentation about why it's being made, and some consideration of how it's impacting data model users and operations.

interface SimpleMessage {
  type: "simple-message";
  pattern: Pattern;
}

interface Matcher {
  selectors: Expression[];
  variants: Variant[];
}

interface ComplexMessage {
  type: "complex-message";

>   declarations?: Declaration[];

Reviewer 2: This should be `declarations: Declaration[]` per #633

  body: Pattern | Matcher;
}
```

Each message _declaration_ is represented by a `Declaration`,
which connects the `name` of a _variable_
with its _expression_ `value`.
The `name` does not include the `$` sigil of the _variable_,

> or the starting and ending `|` of the literals.

Reviewer 2: I'm confused about this -- since this part is describing what appears on the left-hand side of a declaration, where do literals come in?

The `name` of an `InputDeclaration` MUST be the same
as the `name` in the `VariableRef` of its `VariableExpression` `value`.

> An `ReservedStatement` represents a statement not supported by the implementation.

Reviewer 3: Why this rename? Again, I would expect some argumentation here.

  value: Expression;
}

interface ReservedStatement {
  type: "reserved-statement";
  keyword: string;
  body?: string;
  expressions: Expression[];
}
```

> In a `Matcher` the `keys` and `value` of each _variant_ are represented as an array of `Variant`. \

Reviewer 2: ```suggestion
In a `Matcher`,
the `keys` and `value` of each _variant_ are represented as an array of `Variant`. 
```

(Use semantic line breaks; remove stray backslash)

interface CatchallKey {
  type: "*";

>   value: Pattern;

Reviewer 3: This seems like a misunderstanding of this data model entry.

Reviewer 2: Specifically: I think it should be `string` because the intention here is for it to be a different string representing the catchall key in some syntax other than MF2. And it should be optional.

}
```

## Patterns

Each `Pattern` contains a linear sequence of text and placeholders corresponding

> to potential output of a message.

Reviewer 2: ```suggestion
to the potential output of a message.
```

Each element of the `Pattern` MUST either be a non-empty string, an `Expression`,

> or a `Markup` object. \

Reviewer 2: ```suggestion
or a `Markup`.
```
Delete "object" for consistency with the rest of the sentence (it doesn't say "an `Expression` object").

And, I don't think the backslashes are necessary? (Here and elsewhere.)

When they are used as the `body` of an `Expression`,
they represent _expression_ values with no _annotation_.

`Literal` represents all literal values, both numeric (_unquoted_) and string (_quoted_).
The presence or absence of quotes (`'|'` in MessageFormat 2) is not preserved by the data model.
The `value` of `StringLiteral` is the "cooked" value (i.e. escape sequences are processed).

> The `value` of `NumberLiteral` is the numeric value after parsing (for example the trailing zeros are not preserved).

Reviewer 2: ```suggestion
The `value` of `NumberLiteral` is the numeric value after parsing,
(for example the trailing zeros are not preserved).
```

Use semantic line breaks.

More substantively: making this the "parsed" numeric value makes it hard to test the roundtrip property (maybe I missed an issue where this was discussed?)

In a `VariableRef`, the `name` does not include the initial `$` of the _variable_.

```ts
interface VariableRef {
  type: "variable";
  name: string;
}

type Literal = StringLiteral | NumberLiteral

interface StringLiteral {
  type: "string-literal";
  value: string;
}

interface NumberLiteral {
  type: "number-literal";
  value: number;
}
```

A `FunctionAnnotation` represents a _function_ _annotation_.
The `identifier` does not include the `:` starting sigil.

Each _option_ is represented by an `Option`.

```ts
interface Option {
  identifier: Identifier;
  value: Literal | VariableRef;
}
```

An `PrivateUseAnnotation` represents a _private-use annotation_,
not supported by the implementation.
These are annotations that might be supported by some implementations,
and are guaranteed to not be in conflict with future standard development.

An `ReservedAnnotation` represents a _reserved annotation_.

> not supported by the implementations.

Reviewer 2: ```suggestion
not supported by the implementation.
```

(Consistency with the previous para)

> These are annotations that might be defined in the future version of the standars.

Reviewer 2: ```suggestion
These are annotations that might be defined in the future version of the standard.
```

> this data model.

When parsing the syntax of a _message_ that includes a _private-use annotation_
or _reserved annotation_ supported by the implementation,
the implementation SHOULD represent it in the data model
using an interface appropriate for the semantics and meaning
that the implementation attaches to that _annotation_.

## Identifiers

> An `Identifier` is used to identify functions, options, attributes, markup,
> and supporting namespaces to avoid conflict between various proprietary extensions. \

Reviewer 2: ```suggestion
An `Identifier` is used to identify functions, options, attributes, and markup.
It also includes a namespace, to avoid conflict between various proprietary extensions.
```
